Framework/Algorithms/src/ReflectometryReductionOneAuto3.cpp

```diff
@@ -0,0 +1,1148 @@
+
+void validate(const std::string &method) {
+  if (!CorrectionMethod::OPTION_NAME.count(method))
+    throw std::invalid_argument("Unsupported polarization correction method: " +
+                                method);
+}
+} // namespace CorrectionMethod
+
+std::vector<std::string> workspaceNamesInGroup(std::string const &groupName) {
+  auto group =
+      AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(groupName);
+  return group->getNames();
+}
+
+std::string vectorToString(std::vector<std::string> const &vec) {
+  std::string result;
+  for (auto item : vec) {
+    if (!result.empty())
+      result += ",";
+    result += item;
+  }
+  return result;
+}
+
+void removeAllWorkspacesFromGroup(std::string const &groupName) {
+  auto group =
+      AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(groupName);
+  group->removeAll();
+}
+
+void removeWorkspacesFromADS(std::vector<std::string> const &workspaceNames) {
+  for (auto workspaceName : workspaceNames)
+    AnalysisDataService::Instance().remove(workspaceName);
+}
+} // namespace
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(ReflectometryReductionOneAuto3)
+
+namespace {
+
+std::string const OUTPUT_WORKSPACE_BINNED_DEFAULT_PREFIX("IvsQ_binned");
+std::string const OUTPUT_WORKSPACE_DEFAULT_PREFIX("IvsQ");
+std::string const OUTPUT_WORKSPACE_WAVELENGTH_DEFAULT_PREFIX("IvsLam");
+} // namespace
+
+//----------------------------------------------------------------------------------------------
+
+/// Algorithm's name for identification. @see Algorithm::name
+const std::string ReflectometryReductionOneAuto3::name() const {
+  return "ReflectometryReductionOneAuto";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int ReflectometryReductionOneAuto3::version() const { return 3; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string ReflectometryReductionOneAuto3::category() const {
+  return "Reflectometry\\ISIS";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string ReflectometryReductionOneAuto3::summary() const {
+  return "Reduces a single TOF/Lambda reflectometry run into a mod Q vs I/I0 "
+         "workspace attempting to pick instrument parameters for missing "
+         "properties";
+}
+
+/** Validate transmission runs
+ *
+ * @return :: result of the validation as a map
+ */
+std::map<std::string, std::string>
+ReflectometryReductionOneAuto3::validateInputs() {
+  std::map<std::string, std::string> results;
+
+  // Validate transmission runs only if our input workspace is a group
+  if (!checkGroups())
+    return results;
+
+  auto group = AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(
+      getPropertyValue("InputWorkspace"));
+  if (!group)
+    return results;
+
+  // First transmission run
+  const std::string firstStr = getPropertyValue("FirstTransmissionRun");
```

Comment:
I think the checks for the `FirstTransmissionRun` and `SecondTransmissionRun` are exactly the same - can they be refactored into a function, maybe `validateTransmissionRun`. It seems it needs only the property name and a reference to the results map and group.

Reply:
This is from v2 so I've added it to #26499